[dynamo][improve-guard-overhead] Skip TENSOR_MATCH guards on parameters for optimizers

anijain2305 opened the pull request:
**1.32x  guard overhead reduction** (1.092 vs vs 0.827 ms) for MegatronBertForCausalLM with 394 params.


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #122812
* #122725
* #122818
* #122769
* #122716
* __->__ #122647
* #122646



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

anijain2305 committed:
[wip][guards][optimizer] Skip TENSOR_MATCH guards on parameteres for optimizers

[ghstack-poisoned]

anijain2305 committed:
Update on "[wip][guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"



One datapoint - **1.32x  guard overhead reduction** (1.092 vs vs 0.827 ms) for MegatronBertForCausalLM with 394 params.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

anijain2305 commented on test/inductor/test_compiled_optimizers.py:
```diff
@@ -84,12 +84,12 @@ class KernelCounts(NamedTuple):
     Adam: KernelCounts(multitensor=2, singletensor=8),
     AdamW: KernelCounts(multitensor=2, singletensor=8),
-    NAdam: KernelCounts(multitensor=2, singletensor=12),
+    NAdam: KernelCounts(multitensor=2, singletensor=8),
```
Tracing optim.state first instead of tracing optim.param_groups changes input ordering for the AOT graphs. This affects the input graph for inductor, which leads to different fusions.

mlazos replied:
Yeah this is great! I was wondering why NAdam was so bad tbh

anijain2305 committed:
Update on "[guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"



One datapoint - **1.32x  guard overhead reduction** (1.092 vs vs 0.827 ms) for MegatronBertForCausalLM with 394 params.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"



One datapoint - **1.32x  guard overhead reduction** (1.092 vs vs 0.827 ms) for MegatronBertForCausalLM with 394 params.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

anijain2305 committed:
Update on "[guards][optimizer] Skip TENSOR_MATCH guards on parameters for optimizers"



One datapoint - **1.32x  guard overhead reduction** (1.092 vs vs 0.827 ms) for MegatronBertForCausalLM with 394 params.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

mlazos commented:
@anijain2305 are those numbers with cpp guards enabled? just curious

anijain2305 commented:
No, just python guards.

anijain2305 committed:
Update on "[dynamo][improve-guard-overhead] Skip TENSOR_MATCH guards on parameters for optimizers"

**1.32x  guard overhead reduction** (1.092 vs vs 0.827 ms) for MegatronBertForCausalLM with 394 params.





cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge
